node/sc/mainbridge_test.go

```diff
@@ -116,3 +117,79 @@ func TestMainBridge_basic(t *testing.T) {
 
 	//TODO more test
 }
+
+// TestMainBridge_removePeer tests correct removal of a peer from `MainBridge.peers`.
+func TestMainBridge_removePeer(t *testing.T) {
+	// Create a MainBridge (it may have 0 peers)
+	mBridge := testNewMainBridge(t)
+	defer mBridge.chainDB.Close()
+
+	// Prepare a bridgePeer to be added and removed
+	nodeID := "0x1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d"
+	peer := p2p.NewPeer(discover.MustHexID(nodeID), "name", []p2p.Cap{})
+	bridgePeer := mBridge.newPeer(1, peer, &p2p.MsgPipeRW{})
+
+	// Add the bridgePeer
+	mBridge.peers.Register(bridgePeer)
+	peerNum := mBridge.peers.Len()
+
+	// Try to remove a non-registered bridgePeer, and nothing happen
+	mBridge.removePeer("0x11111111")
+	assert.Equal(t, peerNum, mBridge.peers.Len())
+
+	// Remove the registered bridgePeer
+	mBridge.removePeer(bridgePeer.GetID())
+	assert.Equal(t, peerNum-1, mBridge.peers.Len())
+}
+
+// TestMainBridge_handleMsg fails when a bridgePeer fails to read a message or reads a too long message.
+func TestMainBridge_handleMsg(t *testing.T) {
+	// Create a MainBridge
+	mBridge := testNewMainBridge(t)
+	defer mBridge.chainDB.Close()
+
+	// Elements for a bridgePeer
+	nodeID := "0x1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d1d"
+	peer := p2p.NewPeer(discover.MustHexID(nodeID), "name", []p2p.Cap{})
+	pipe1, pipe2 := p2p.MsgPipe()
+
+	// bridgePeer will receive a message through rw1
+	bridgePeer := newBridgePeer(testProtocolVersion, peer, pipe1)
+
+	// Case1. Send a valid message and handle it successfully
+	{
+		data := "valid message"
+		go func() {
+			if err := p2p.Send(pipe2, StatusMsg, data); err != nil {
```

Comment:
Should this routine be executed with goroutine?
There is no guarantee that this code will run before `handleMsg` routine caused by goroutine.

Reply:
1. It needs to be a goroutine because `p2p.Send` is a blocking function.
2. `mBridge.handleMsg` will wait until it receives a channel message from another pipe.